Guest Blog: Ringo De Smet from Cumundi (#2747)
pulumi/docs: content/blog/cumundi-guest-post/index.md

---
title: "Refactoring Pulumi Code with `aliases`"
date: 2020-03-26
meta_desc: "Cumundi uses Pulumi to build infrastructure components for their customers"
meta_image: cumundi-pulumi.png
authors:
    - ringo-de-smet
tags:
    - Test Driven Developmnet
---

**Guest Article:** [Ringo De Smet](https://www.linkedin.com/in/ringodesmet/), Founder of [Cumundi](https://www.cumundi.cloud), standardizes on Pulumi for writing infrastructure as code libraries to deliver reusable building blocks. Pulumi enables him to iterate through the build-test-release cycle of these building blocks rapidly.

> Cumundi helps companies adopt the cloud more integrally. Companies are rebuilding the same 75% of the underlying infrastructure. But the shortage of people and time has led to limited investment in integrating all the non-functional requirements needed to run applications in the cloud.

Reviewer: Replace with: "standardizes on Pulumi for writing infrastructure as reusable code libraries for his customers. Pulumi enables him to rapidly iterate through the build-test-release cycle of these building blocks.

Cumundi helps companies adopt cloud infrastructure in a more integral way. It has found that 75% of companies’ needs are covered by ‘vanilla’ infrastructure patterns. However, due to a shortage of people and time, there has been limited investment to take full advantage of cloud-native configurations - which can lead to inefficiency, poor performance, and higher costs."

<!--more-->

## How Pulumi Code Keeps Up with Changing Requirements

> At Cumundi, we build reusable libraries for our customers to set up their infrastructure integrated with best practices. These best practices span the range of non-functional requirements, which are most of the time not taken up with the same focus as application feature development. Where these non-functional requirements can be integrated into infrastructure setups, we include these in our code libraries. Cumundi focuses on all three major cloud providers to provide application-focused infrastructure libraries.

Reviewer: ```suggestion
At Cumundi, we build reusable libraries for our customers to set up their infrastructure while following best practices. These best practices include a range of "non-functional requirements" - infrastructure patterns - which often receive less focus than application features. Where feasible, Cumundi encapsulates these non-functional requirements into infrastructure code libraries. This enables Cumundi to serve clients using all major clouds while keeping infrastructure libraries clean and modular.
```

Author: changed it a bit to keep the intent of application focus more clear

> For our internal infrastructure, it is not surprising that we also use Pulumi. At this early phase of our company (founded January 1st!), it is hard to foresee how we will evolve. With Pulumi, we write code in a modern programming language. A good coding practice is not to [optimize prematurely](http://wiki.c2.com/?PrematureOptimization) because the future is unpredictable. Another coding best practice is [red-green-refactor](https://en.wikipedia.org/wiki/Test-driven_development) or formally known as Test Driven Development. You take the current code, write a test for the new requirement, which initially fails (red). Next, you implement the code in the most straightforward way to make the test succeed (green) and you complete the cycle by refactoring the code to keep the design in a proper shape.

Reviewer: ```suggestion
It may come as no surprise that we also use Pulumi to manage Cumundi's own infrastructure. As a young company (founded January 1st!), it is hard to foresee how our infrastructure needs will evolve. With Pulumi, we write code in a modern programming language, and, good coding practice is not to [optimize prematurely](http://wiki.c2.com/?PrematureOptimization) because the future is unpredictable. Another coding best practice is [red-green-refactor](https://en.wikipedia.org/wiki/Test-driven_development) or formally known as Test-Driven Development. You take the current code, write a test for the new requirement, which initially fails (red). Next, you implement the code in the most straightforward way to make the test succeed (green) and you complete the cycle by refactoring the code to keep the design in proper shape.
```

This blog post demonstrates a TDD cycle for Pulumi code with a reduced version of the code used to configure the internal infrastructure we provide for each customer. This example focuses on one specific Pulumi resource property: [`aliases`](https://www.pulumi.com/docs/intro/concepts/programming-model/#aliases)

> All the code is [published here](https://github.com/cumundi/pulumi-refactoring-aliases) if you want to follow along with a full project setup. Every step described here is committed as a separate branch with the starting point on `master`, the default branch.

Reviewer 2: I am not a fan of "published here" - I'd rather we said, "publishing in GitHub" - here isn't a very accessible word

Author: fixed

### The Starting Point

Every Pulumi project starts empty, so ours is no exception. This is how our main Pulumi code file looks for now.

```ts
import * as pulumi from "@pulumi/pulumi";

const config = new pulumi.Config('gitlab')
```

Applying the above code does nothing, which you can see in the *Resources* tab on the hosted Pulumi platform:

![Starting point](./start.png)

### Adding the Git Repository

> We created a separate repository for each of our customers:

Reviewer 2: ```suggestion
We create a separate repository for each of our customers:
```

```ts
const secondCustomer = new gitlab.Project("SecondCustomer",
    {
        ... // properties here
    }
)

const secondCustomerProject = new gcp.organizations.Project("SecondCustomer",
    {
        projectId: 'second-customer',
        name: 'Second Customer Infrastructure'
    }
)

const serviceAccountSecondCustomer = new gcp.serviceAccount.Account("ServiceAccountSecondCustomer",
    {
        accountId: 'secondcustomer',
        displayName: 'Service Account for Second Customer project',
        project: secondCustomerProject.projectId
    }
);

const serviceAccountSecondCustomerKey = new gcp.serviceAccount.Key("ServiceAccountSecondCustomerKey",
    {
        serviceAccountId: serviceAccountSecondCustomer.email
    }
);

const secondCustomerGitlabCIVariable = new gitlab.ProjectVariable("SecondCustomerGCPAccess",
    {
        project: secondCustomer.id,
        variableType: 'file',
        key: "GOOGLE_APPLICATION_CREDENTIALS",
        value: serviceAccountSecondCustomerKey.privateKey.apply(value => Buffer.from(value, 'base64').toString('utf-8')),
        protected: true,
        environmentScope: "*"
    }
)
```

> These resources are all added with the `Stack` as their parent in the resource view, as seen below. This is getting clumsy.

Reviewer: ```suggestion
These resources are all added with the `Stack` as their parent in the resource view, as seen below. With many resources, this view can get complicated quickly.
```

![Step 2](./step2.png)

> Visualization goes a long way, but a flat list of resources does not clearly show what belongs together.

Reviewer: ```suggestion
Visualization goes a long way, but a flat list of resources does not clearly show what belongs together. Fortunately, there is a better way to keep these resources organized.
```

> ### Find Relationships between Resources

Reviewer: ```suggestion
### Finding Relationships between Resources
```

In the current state of the code, we created a Google Cloud service account and a key. Since a key can't be created without a service account, we also created a Gitlab CI project variable for the Gitlab project of the second customer. Two cases of a parent-child relationship.

Although we have this relationship between our resources, the Pulumi state graph doesn't display it. How can we change this without affecting the real resources on Gitlab and Google Cloud?

> We can pass [`CustomResourceOptions`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/pulumi/#CustomResourceOptions) as the last argument to every Pulumi resource that we want to create. We use `parent` and `aliases` for refactoring.

Reviewer: ```suggestion
We can pass [`CustomResourceOptions`](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/pulumi/#CustomResourceOptions) as the last argument to every Pulumi resource that we want to create and use `parent` and `aliases` for refactoring.
```
